### PR TITLE
Enable Server Side rendering for logged-out Plugins section

### PR DESCRIPTION
**Pull request by claudiucelfilip**
#### Proposed Changes

More info: pau2Xa-4j7-p2

This builds on top of #66169 and adds Server Side Rendering to the logged-out Plugins section.

Changes for this include:
* prefetched data inside route middleware so they're ready for when the components are rendered
* hydrate the client site with server fetched data
* adjust and harden client-specific functions
* force the Swipe layout for the Category links so they get rendered

#### Testing Instructions

* if you're running locally, make sure to run `yarn install` beforehand
* make sure you're logged out 
* disable JavaScript
* all links should be accessible:

* `/plugins`
Screenshot on 2022-08-22 at 15-00-31

* `/plugins?s=seo`
Screenshot on 2022-08-22 at 14-58-31

* `/plugins/browse/seo`
Screenshot on 2022-08-22 at 14-58-58

* `/plugins/wordpress-seo-premium`

Screenshot on 2022-08-22 at 16-37-48

* clicking on the `Visit plans` CTA should send you to https://wordpress.com/pricing/

* all the links that should be inaccessible and should show the loading placeholder logo (enabling JavaScript would redirect to log in)
  * `/plugins/manage`
  * `/plugins/upload`
  * `/plugins/setup`
  * `/plugins/active`
  * `/plugins/inactive`
  * `/plugins/updates`
  * `/plugins/wordpress-seo-premium/eligibility`
  * any link which has a `SITE_ID` in it (ex. `/plugins/browse/seo/[SITE_ID]`)

 Screenshot on 2022-08-22 at 15-02-47
 

* enable JavaScript and log in
* visit the same links
* everything should work as before

#### Pre-merge Checklist

- [x] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [x] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [x] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?

Related to #

### CODE REVIEW

**Review by tyxla — COMMENTED** [review]
Just saw this PR, thanks for working on this, pretty exciting 👏 

In my quick tests, I noticed that for logged-in users, we shortly display the logged-out masterbar. I haven't investigated, but this seems to hint that we're trying to SSR for logged-in users, and I don't think we should be doing that. Is this something you're still working on?

**Comment by claudiucelfilip** [comment]
> this seems to hint that we're trying to SSR for logged-in users, and I don't think we should be doing that. Is this something you're still working on?

@tyxla Thanks for catching this. The SSR should now be disabled for logged-in users. It's also disabled if the `site` is present, and the client side should handle redirecting to log-in page.

File: client/lib/plugins/sanitize-section-content.js
Comment (tyxla) [inline client/lib/plugins/sanitize-section-content.js]: I'm not convinced this is always the right level to disable browser-only behavior to enable SSR. 
For example, in this instance, are we just going to leave the content unsanitized, and potentially vulnerable?

Comment (claudiucelfilip) [reply client/lib/plugins/sanitize-section-content.js]: @tyxla I pushed a [workaround](https://github.com/Automattic/wp-calypso/pull/66344/commits/95b6357cd408ad7525844407ba384c3773fd2453#diff-33ff16937e1a3218ed65c949cc79ec6f88ec3efb06d89d8d1cb1e0f0f49b32ae) based on [htmlparser2](https://github.com/fb55/htmlparser2), [domutils](https://github.com/fb55/domutils) and [dom-serializer](https://github.com/cheeriojs/dom-serializer)

These should load only on the server and replaces the `window` APIs used in content sanitization with server-friendly equivalents.

Comment (tyxla) [reply client/lib/plugins/sanitize-section-content.js]: Thanks for this! I haven't reviewed thoroughly, but given the complexity that the additional boilerplate provides, I wonder if we've tried a more targeted and ready-to-use solution, like [sanitize-html](https://www.npmjs.com/package/sanitize-html) for example?

Comment (claudiucelfilip) [reply client/lib/plugins/sanitize-section-content.js]: `sanitize-html` seems to have been used previously but it was replaced in #19316 with this `sanitizer-section-content`. 
Since then, a few other transforms have been added and the outputs have diverge. Running the `sanitizer-section-content` tests against the `sanitize-html`  results in most tests failing.
 
Even if we configure `sanitize-html` and `sanitizer-section-content` to return similar outputs, I think it'll become trickier to maintain and develop both approaches.

Comment (tyxla) [reply client/lib/plugins/sanitize-section-content.js]: Okay, I see. I know we're going in circles, but are you sure that sanitizing this on the endpoints won't be easier than what's being suggested here? I understand they're 4 endpoints, but it sounds like we could reuse the functionality.

Comment (claudiucelfilip) [reply client/lib/plugins/sanitize-section-content.js]: I don't think it'd be any easier because one of the endpoints is on `api.wordpress.org`. 
I'm not too familiar with that codebase I doubt we'll be able to push the changes. Even so, we'd probably need to implement and then manage the content sanitization on both platforms.

Comment (tyxla) [reply client/lib/plugins/sanitize-section-content.js]: Thanks for confirming. That does add additional challenges, indeed. 

Let's take what you have as a solution. What remains to be done in order for this PR to be ready for review? I see some tests are failing.

Comment (claudiucelfilip) [reply client/lib/plugins/sanitize-section-content.js]: The tests should be addressed: the unit tests needed the magnificent locale mocked, and the e2e was caused by the transition between SSR and client render defocusing the search input.

I think there should be enough here ready for review. 
From my testing, the routes work as expected and the content renders properly on both mediums. 

There's still some flickering mostly caused by a few components rendering differently based on breakpoint but it might not be worth addressing them in this already massive PR.

File: client/data/marketplace/use-es-query.ts
Comment (claudiucelfilip) [inline client/data/marketplace/use-es-query.ts:111]: Needed to extract the `react-query` cache key and the fetch function pair in order to use them in the server prefetch middleware.

File: client/document/index.jsx
Comment (claudiucelfilip) [inline client/document/index.jsx:76]: Similar to the redux store, this dehydrates the `react-query` state and will be used on the client side.

File: client/lib/plugins/sanitize-section-content.js
Comment (claudiucelfilip) [inline client/lib/plugins/sanitize-section-content.js:14]: This allows us to insert a custom server-friendly `DOMParser` in the content sanitizer from the node controller.

File: client/lib/plugins/server-dom-parser.js
Comment (claudiucelfilip) [inline client/lib/plugins/server-dom-parser.js:125]: This needs to match all the interfaces needed in the plugin content sanitizer.

File: client/my-sites/plugins/plugin-details.jsx
Comment (claudiucelfilip) [inline client/my-sites/plugins/plugin-details.jsx:249]: The server never calls the `useEffect` hook, and thus never sets the breadcrumbs items. Calling the breadcrumb functionality through `useServerEffect` allows us to only run it on the server and leave the client flow untouched.

File: client/server/render/index.js
Comment (claudiucelfilip) [inline client/server/render/index.js:237]: Adding the`productsList` here avoids the situation where the server fetches the plugin prices in a different currency than the client. It should also improve performance by not having to fetch the `productsList` again on the client side.

File: client/state/query-client.ts
Comment (claudiucelfilip) [inline client/state/query-client.ts:10]: Reusing the same `queryClient` allows us to cache the server data.

File: client/state/query-client.ts
Comment (claudiucelfilip) [inline client/state/query-client.ts:21]: We don't want to cache the `queryClient` for logged in users.

File: client/webpack.config.node.js
Comment (claudiucelfilip) [inline client/webpack.config.node.js:166]: This seems to have been missing for the node build config. Without this, package components (such as Search) break because they use `__i18n_text_domain__` for localization.

File: package.json
Comment (claudiucelfilip) [inline package.json:193]: `htmlparser2`, `domutils`, `dom-serializer` are needed for the server-friendly DOMParser equivalent for the plugin content sanitizer.

File: package.json
Comment (claudiucelfilip) [inline package.json:298]: This was added to address the package deduping errors.

File: packages/i18n-utils/src/locale-context.tsx
Comment (claudiucelfilip) [inline packages/i18n-utils/src/locale-context.tsx:83]: Needed to add `typeof` to avoid SSR errors.

File: test/e2e/specs/plugins/plugins__search.ts
Comment (claudiucelfilip) [inline test/e2e/specs/plugins/plugins__search.ts]: This is needed because the search input gets defocused when the SSR version of the content gets switched with the client one.

File: client/my-sites/plugins/categories/use-categories.tsx
Comment (claudiucelfilip) [inline client/my-sites/plugins/categories/use-categories.tsx]: Needed to extract this into a function in order to construct the categories server prefetch query.

File: client/server/render/index.js
Comment (noahtallen) [reply client/server/render/index.js:237]: By the way, you can now add this below on line 258 to improve logged out performance -- assuming that caching `productsList` in the state allows us to avoid an extra network request on the server. We improved state caching so that it doesn't operate for logged-in requests now. 

(Obviously only do this if the data in `productsList` matches 1:1 with the route after accounting for locale and everything!)

File: test/e2e/specs/plugins/plugins__search.ts
Comment (worldomonation) [reply test/e2e/specs/plugins/plugins__search.ts]: **question**: Instead of adding a floating `waitForLoadState('networkidle')` in the spec itself, would it make sense to apply this wait directly in the `pluginsPage.visit` method? 

That way, every call to `pluginsPage.visit` will benefit from the wait and help eliminate potential race conditions _and_ it would also reduce instances of "raw" calls to Playwright methods in specs.

**Review by worldomonation — COMMENTED** [review]
Left a minor question.

**Comment by mreishus** [comment]
@claudiucelfilip do you want us to focus on this one or #66169 first (that is, which do you plan to merge first?)

**Comment by mreishus** [comment]
Having so much of the plugins section work with JS off is impressive!


The only things I were able to notice were:
* `/es/plugins` being in all English as reported in [66169](https://github.com/Automattic/wp-calypso/pull/66169#issuecomment-1227668022)
* While `http://calypso.localhost:3000/plugins?s=seo` works correctly, visiting `http://calypso.localhost:3000/plugins`, typing "seo" in the box and pressing enter with JS off, sends me to `http://calypso.localhost:3000/?`. Not sure if this is fixable and seems pretty minor overall.

https://user-images.githubusercontent.com/937354/186758106-bd7b235a-a75c-48a3-9b51-1ecbf58301a9.mp4

File: client/my-sites/plugins/categories/use-categories.tsx
Comment (lsl) [inline client/my-sites/plugins/categories/use-categories.tsx]: regression?

File: client/server/render/index.js
Comment (claudiucelfilip) [reply client/server/render/index.js:237]: Thanks @noahtallen for the improvements to the state caching. I added `productsList` and it seems to work well.

**Review by niranjan-uma-shankar — COMMENTED** [review]
I ran into the same issue noted by @mreishus [here](https://github.com/Automattic/wp-calypso/pull/66344#issuecomment-1227704859):

> While http://calypso.localhost:3000/plugins?s=seo works correctly, visiting http://calypso.localhost:3000/plugins, typing "seo" in the box and pressing enter with JS off, sends me to http://calypso.localhost:3000/?. Not sure if this is fixable and seems pretty minor overall.

Would this impact SEO? If not, probably not needed to be fixed urgently. 


Secondly, I tried loading the plugins page in a couple of different locales, closed all incognito windows, and opened a fresh incognito window after disabling JS. You can see in the video below how I'm getting the page rendered in three different locales - `ja`, `es`, and `en` - when I would've expected the page to be only in `en` (notice that the URL does not have a locale attached). I guess this is to do with cache not getting busted for a new locale, would it be a problem for SEO purposes?

https://user-images.githubusercontent.com/1269602/186861361-cf37229c-8b22-421f-aa1b-b9bba74075fa.mov


For the SSR layout, is there anything from a SEO perspective that needs to be handled? Like robots.txt etc. I don't know much on this front, maybe good to get a sign-off from @dkd903 .

File: client/my-sites/plugins/categories/use-categories.tsx
Comment (claudiucelfilip) [reply client/my-sites/plugins/categories/use-categories.tsx]: Indeed. Thanks for catching it!

**Comment by claudiucelfilip** [comment]
> do you want us to focus on this one or #66169 first (that is, which do you plan to merge first?)

@mreishus Focusing first on #66169 sounds good but I'd follow up with this PR as soon as possible.

**Comment by claudiucelfilip** [comment]
> > While http://calypso.localhost:3000/plugins?s=seo works correctly, visiting http://calypso.localhost:3000/plugins, typing "seo" in the box and pressing enter with JS off, sends me to http://calypso.localhost:3000/?. Not sure if this is fixable and seems pretty minor overall.
> 
> Would this impact SEO? If not, probably not needed to be fixed urgently.

I agree, and even if it's a good SEO opportunity I would rather we follow up in a separate PR.

> Secondly, I tried loading the plugins page in a couple of different locales, closed all incognito windows, and opened a fresh incognito window after disabling JS. 

Thanks for catching this! 
It was caused by the `defaultI18n` module not resetting properly for English locales in the `ssrSetupLocale` middleware. I pushed the fix this.

File: client/my-sites/plugins/categories/use-categories.tsx
Comment (claudiucelfilip) [inline client/my-sites/plugins/categories/use-categories.tsx]: Needed to use this `__` in order to be able to translate the categories in the prefetch middleware (before the provider and components are initialized).

File: client/my-sites/plugins/plugins-navigation-header/index.jsx
Comment (lsl) [inline client/my-sites/plugins/plugins-navigation-header/index.jsx]: Fixed these in https://github.com/Automattic/wp-calypso/pull/67140, you'll still need the `setBreadcrumbs` change.

File: client/lib/plugins/sanitize-section-content.js
Comment (claudiucelfilip) [inline client/lib/plugins/sanitize-section-content.js]: This was needed because the server-friendly parser cannot construct the URL object within the `a` element.

Comment (lsl) [reply client/lib/plugins/sanitize-section-content.js]: Probably want a try catch around this just to be safe. Relative paths in any links will throw here.

Comment (claudiucelfilip) [reply client/lib/plugins/sanitize-section-content.js]: Thanks for catching that! I pushed a fix.

**Comment by markbiek** [comment]
I'm not familiar enough to know if there's potential for any unintended side-effects with any of this but I wanted to pop in and say that everything in the test plan worked as expected 🎉

**Review by niranjan-uma-shankar — APPROVED** [review]
Test plan worked for me, LGTM.

I noticed a minor bug in [breadcrumb translation](https://github.com/Automattic/wp-calypso/pull/67140#issuecomment-1236962444) but no reason for that issue to block this PR.

**Comment by noahtallen** [comment]
@claudiucelfilip we had to revert this PR today because it was causing performance issues on trunk :(

**Pull request merged**
